kernel/src/actions/mod.rs

```diff
@@ -318,44 +328,85 @@ impl Protocol {
     }
 }
 
-// given unparsed `table_features`, parse and check if they are subset of `supported_features`
-pub(crate) fn ensure_supported_features<T>(
-    table_features: &[String],
-    supported_features: &HashSet<T>,
-) -> DeltaResult<()>
+#[inline]
+fn create_feature_error<T>(
+    unsupported: Vec<T>,
+    unsupported_or_unknown: &str,
+    features: HashSet<T>,
+) -> Error
 where
-    <T as FromStr>::Err: Display,
-    T: Debug + FromStr + Hash + Eq,
+    T: ToString + Debug,
 {
-    let error = |unsupported, unsupported_or_unknown| {
-        let supported = supported_features.iter().collect::<Vec<_>>();
-        let features_type = type_name::<T>()
-            .rsplit("::")
-            .next()
-            .unwrap_or("table features");
-        Error::Unsupported(format!(
-            "{} {} {:?}. Supported {} are {:?}",
-            unsupported_or_unknown, features_type, unsupported, features_type, supported
-        ))
-    };
-    let parsed_features: HashSet<T> = table_features
+    let unsupported = unsupported
         .iter()
-        .map(|s| T::from_str(s).map_err(|_| error(vec![s.to_string()], "Unknown")))
-        .collect::<Result<_, Error>>()?;
+        .map(ToString::to_string)
+        .collect::<Vec<_>>();
 
-    // check that parsed features are a subset of supported features
+    Error::Unsupported(format!(
+        "{} reader features {:?}. Supported reader features are {:?}",
+        unsupported_or_unknown, unsupported, features
+    ))
+}
+
+pub(crate) trait UnknownFeature: Clone + Hash + Eq + Display + Debug {
+    fn is_unknown(&self) -> bool;
+    fn build_unknown(feature: String) -> Self;
+}
+
+impl UnknownFeature for ReaderFeatures {
+    fn is_unknown(&self) -> bool {
+        matches!(self, ReaderFeatures::Unknown(_))
+    }
+
+    fn build_unknown(feature: String) -> Self {
```

Comment:
i think we usually take something like `impl ToString` for these helper methods.